src/diagonal.works/b6/api/functions/collections.go

```diff
@@ -341,3 +342,155 @@ var _ api.Collection = &flattenCollection{}
+	start, end interface{}
+}
+
+func bucketRange(b bucket) interface{} {
+	if b.start == b.end || b.end == nil {
+		return b.start
+	} else {
+		return fmt.Sprint(b.start) + "-" + fmt.Sprint(b.end)
+	}
+}
+
+type bucketFn func(interface{}, bucket) bool // Indicates whether given value can be bucketed.
+
+func bucketIndex(v interface{}, b []bucket, fn bucketFn) (int, error) {
+	for i, bucket := range b {
+		if fn(v, bucket) {
+			return i, nil
+		}
+	}
+
+	return -1, fmt.Errorf("can't bucket %v", v)
+}
+
+type kv struct {
+	key   interface{}
+	value int // count
+}
+
+func anyBuckets(kvs []kv) ([]bucket, error) {
+	sort.Slice(kvs, func(i, j int) bool {
+		return kvs[i].value > kvs[j].value
+	})
+
+	var buckets []bucket
+	for i, kv := range kvs {
+		if i == maxBuckets {
+			break
+		}
+
+		if i == maxBuckets-1 && len(kvs) > maxBuckets {
+			buckets = append(buckets, bucket{start: "other"})
+		} else {
+			buckets = append(buckets, bucket{start: kv.key})
+		}
+	}
+
+	return buckets, nil
+}
+
+func anyBucketFn(v interface{}, b bucket) bool {
+	return b.start == v || b.start == "other"
+}
+
+func uniformBuckets(kvs []kv) ([]bucket, error) {
+	sort.Slice(kvs, func(i, j int) bool {
+		iKey, _ := api.ToFloat64(kvs[i].key) // a: not dealing w errors here either, relying on ConvertibleToFloat64 check in histogram
+		jKey, _ := api.ToFloat64(kvs[j].key)
+		return iKey > jKey
+	})
+
+	var buckets []bucket
+	if (len(kvs)) <= maxBuckets {
+		for _, kv := range kvs {
+			buckets = append(buckets, bucket{start: kv.key, end: kv.key})
+		}
+	} else {
+		for len(kvs) > 0 {
+			bucket_size := len(kvs) / (maxBuckets - len(buckets))
+			buckets = append(buckets, bucket{start: kvs[0].key, end: kvs[bucket_size-1].key})
+			kvs = kvs[bucket_size:]
+		}
+	}
+
+	return buckets, nil
+}
+
+func numericBucketFn(v interface{}, b bucket) bool {
+	start, _ := api.ToFloat64(b.start) // a: not propagating errors / felt duplicate and ugly / relying on ConvertibleToFloat64 in histogram(), but lmk if u want it!
+	end, _ := api.ToFloat64(b.end)
+	float, _ := api.ToFloat64(v)
+
+	return float >= start && float <= end
+}
+
+func countvalues(c api.Collection) ([]kv, error) { // a: this is exactly like count-values but if i reuse either for either it s extra work and makes no sense :( what do i do
+	m := make(map[interface{}]int)
+
+	i := c.Begin()
+	for {
+		ok, err := i.Next()
+		if err != nil {
+			return nil, err
+		}
+		if !ok {
+			break
+		}
+
+		m[i.Value()]++
+	}
+
+	var kvs []kv
+	for k, v := range m {
+		kvs = append(kvs, kv{k, v})
+	}
+
+	return kvs, nil
+}
+
+func histogram(co *api.Context, c api.Collection) (api.Collection, error) {
+	values, err := countvalues(c)
+	if err != nil {
+		return nil, err
+	}
+
+	var buckets []bucket
+	var fn bucketFn
+	// Checking only first, assuming all elements in collection have the same type. TODO: enforce/check that somewhere else.
+	if api.ConvertibleToFloat64(values[0].key) { // a: do i need to check empty / collection before going in
```

Comment:
Yeah - if you're exposing this from the API, you should handle both empty and nil collections gracefully (I think nil can return an error, but empty should probably show an empty histogram without error).